### ISSUE TITLE
Unexpected compile error when accessing prop via super.prop

### ISSUE THREAD

**Issue by LeviticusMB** [microsoft/TypeScript#59892]
### 🔎 Search Terms

"not accessible in the child class via super"

### 🕗 Version & Regression Information

- This changed between versions 5.2 and 5.3.
- This changed in commit or PR _______
- This is the behavior in every version I tried, and I reviewed the FAQ for entries about _________
- I was unable to test this on prior versions because _______

### ⏯ Playground Link

https://github.com/LeviticusMB/demo-ts-super

### 💻 Code

```ts
class URI extends URL {
    override readonly href!: string;
    override readonly origin!: string;
    override readonly protocol!: string;
    override readonly username!: '';
    override readonly password!: '';

    constructor(url: string | URL, base?: string | URL) {
        super(url, base);
        this.username = this.password = '';

        Object.defineProperty(this, 'href',     { get: () => super.href     });
        Object.defineProperty(this, 'origin',   { get: () => super.origin   });
        Object.defineProperty(this, 'protocol', { get: () => super.protocol });
        Object.defineProperty(this, 'username', { get: () => super.username });
        Object.defineProperty(this, 'password', { get: () => super.password });
    }

    set _href(href: string) {
        super.href = href;
    }
}

function guard(fn: () => void) {
    try {
        fn();
    } catch (e) {
        console.warn(`${fn} threw ${e}`);
    }
}

console.assert(false, 'console.assert works!')

// Username/password will be removed
const uri = new URI('https://foo:bar@example.com?query#hash');
console.log(uri.toString());
console.assert(uri.href === 'https://example.com/?query#hash', uri.href);
console.assert(uri.username === '', uri.username);
console.assert(uri.password === '', uri.password);

// Read-only properties
// @ts-expect-error readonly
guard(() => uri.username = 'user');
// @ts-expect-error readonly
guard(() => uri.password = 'pass');
// @ts-expect-error readonly
guard(() => uri.protocol = 'ftp:');
console.log(uri.toString());
console.assert(uri.href === 'https://example.com/?query#hash', uri.href);
console.assert(uri.username === '', uri.username);
console.assert(uri.password === '', uri.password);
console.assert(uri.protocol === 'https:', uri.protocol);

// Cannot change href either
// @ts-expect-error readonly
guard(() => uri.href = 'ftp://download.org');
console.log(uri.toString());
console.assert(uri.href === 'https://example.com/?query#hash', uri.href);
console.assert(uri.origin === 'https://example.com', uri.origin);
console.assert(uri.protocol === 'https:', uri.protocol);

// Anything that doesn't mess with the protocol is allowed to change
uri.hostname = 'download.org';
uri.port = '8080';
uri.pathname = 'path';
uri.search = 'params';
uri.hash = 'fragment';
console.log(uri.toString());
console.assert(uri.href === 'https://download.org:8080/path?params#fragment', uri.href);
console.assert(uri.origin === 'https://download.org:8080', uri.origin);
console.assert(uri.host === 'download.org:8080', uri.host);
console.assert(uri.hostname === 'download.org', uri.hostname);

// Magic _href works
uri._href = 'http://wiki.net/docs?search#anchor';
console.log(uri.toString());
console.assert(uri.href === 'http://wiki.net/docs?search#anchor', uri.href);
console.assert(uri.origin === 'http://wiki.net', uri.origin);
console.assert(uri.protocol === 'http:', uri.protocol);

```

### 🙁 Actual behavior

*Class field 'href' defined by the parent class is not accessible in the child class via super.*

### 🙂 Expected behavior

No errors, since the code works just fine.

### Additional information about the issue

I have a Node.js class hierarchy that extends `URL` to perform actions on URLs and URIs, but since each subclass contains the implementation of a given scheme/protocol, i want to restrict write access to a few of the URL properties.

I do this by defining getters with the same name. (Additionally, I have a protected setter for the use case when a subclass actually needs to modify the `href` property for some reasons).

This used to compile just fine, but when I bumped TS a while back, the code no longer compiles.

I reduced the code to a small example which I've provided a link to. The code actually runs just fine on any Node.js version I have ever tried. I've also tried to run the demo in all major browsers, without problems. So I don't understand why TS thinks the code is broken?

**Comment by LeviticusMB** [comment]
I've also checked that it works both with and without "use strict", BTW.

**Comment by jcalz** [comment]
https://github.com/LeviticusMB/demo-ts-super That's not a playground link. Use https://www.typescriptlang.org/play and demo the issue there.

**Comment by LeviticusMB** [comment]
Well yes, I had the demo code ready before I filed the issue.

I'm not able to produce an equivalent set-up (plain Node environment) at the playground. But it does look like the errors go away when the "dom" lib/module is referenced... So the problem might be in `@types/node`.

`lib.dom.d` declares URL as an plain interface. `@types/node` **also** defines URL as an interface, **but** the interface extends a class named `_URL`:

```ts
import { URL as _URL, ... } from "url";
...
interface URL extends _URL {}
```

So apparently URL is not actually an interface in this context, but a class ...? TS issue or DefinitelyTyped?

**Comment by RyanCavanaugh** [comment]
It's declared as a field in `@types/node`; if that's correct then `super.href` genuinely doesn't work; if it's incorrect then it should be changed to
```
        get href(): string;
        set href(s: string );
```

**Comment by Renegade334** [comment]
In the web spec, attributes are implemented as getters/setters, and therefore defined on the prototype, which is why `super.*` works in JS-land. The generated TypeScript DOM library, on the other hand, defines these attributes as instance properties, not getters/setters.

Node.js uses the same attribute getter/setter paradigm for the implementation of its web-compatible APIs. @types/node mirrors the TypeScript library, in that it defines those attributes as instance properties rather than getters/setters.

According to both the TS definitions and the @types/node definitions, those instance properties shouldn't be accessible via `super.*` in either...

The reason this doesn't error in TS when you're using the DOM library is that the web interfaces are defined there as pre-ES6 style constructor functions, on which the compiler doesn't run `super` property access checks. In Node, they're implemented as classes, on which the compiler _does_ run accessibility checks for `super` property access. (@RyanCavanaugh, do you think this discrepancy is something that specifically needs attention?)

As far as @types/node goes, my personal opinion is that its web API definitions should continue to mirror the TypeScript DOM library definitions in terms of the convention for property definitions. If this needs further discussion, it can be raised over at DefinitelyTyped.
